File: DittoDatto-old/packages/shared-types/src/service.ts
import { z } from "zod";
import { IdSchema, PriceSchema, CurrencySchema } from "./common";

export const BookingModeSchema = z.enum([
  "standard", // 1:1 appointments (haircut, massage, etc.)
  "tableReservation", // Capacity-based (restaurants, group therapy)
  "ticketSystem", // Inventory-based (events, classes, workshops)
]);
export type BookingMode = z.infer<typeof BookingModeSchema>;

export const ServiceSchema = z.object({
  id: IdSchema,
  storeId: IdSchema,
  groupId: IdSchema.optional(), // Link to ServiceGroup for config inheritance
  staffId: IdSchema.optional(), // Legacy: single staff. TODO: migrate to assignedStaff[] (see ADR-0006)
  /** Staff UIDs who can perform this service. Replaces staffId for many-to-many. */
  assignedStaff: z.array(IdSchema).default([]),

  /**
   * Resource groups required to perform this service.
   * Empty = no resource needed (current behavior, most services).
   * When set, MercuryEngine checks: staff available AND resource available.
   * References: companies/{companyId}/stores/{storeId}/resourceGroups/{groupId}
   */
  requiredResourceGroupIds: z.array(IdSchema).default([]),

  title: z.string().min(1),
  description: z.string().optional(),

  // Booking Mode: determines how customers book this service
  // - standard: 1:1 time-based appointments
  // - tableReservation: capacity-based (group bookings)
  // - ticketSystem: inventory-based (event tickets)
  bookingMode: BookingModeSchema.default("standard"),

  // Media
  coverImage: z.string().optional(),
  gallery: z.array(z.string()).optional(),

  serviceType: z.array(z.string().min(1)),
  subcategory: z.string().optional(),

  keywords: z.array(z.string().min(1)), // AI Tags <- This one I think should be required, because it is used for AI
  aiDescription: z.string().optional(),
  _embedding: z.array(z.number()).optional(),

  // Time & Pricing
  duration: z.number().int().positive(),
  price: PriceSchema,
  bufferTime: z.number().int().default(0),
  currency: CurrencySchema,

  /**
   * Over-capacity policy (Noona: overbookable enum).
   * - reject: hard limit, "Sorry, no availability" (default)
   * - request: submit as pending_approval, notify business
   * - allow: always accept even if over capacity
   */
  overCapacityPolicy: z.enum(["reject", "request", "allow"]).default("reject"),

  /** Minimum advance notice required to book this service (in minutes). 0 = no minimum. */
  minBookingNoticeMinutes: z.number().int().min(0).max(43200).default(0),
  /** Slot interval for generating bookable time slots (in minutes). Default 15 = :00, :15, :30, :45. */
  slotInterval: z.number().int().min(5).max(120).default(15),

  isActive: z.boolean().default(true),
});
export type Service = z.infer<typeof ServiceSchema>;

File: DittoDatto-old/services/mercury-engine/src/mercury_core/models/service.py
"""Service and ServiceGroup models.

Source: schemas/company-blueprint.surql §1.2–1.3
Zod origin: packages/shared-types/src/service.ts, service-group.ts
"""

from __future__ import annotations

from pydantic import Field

from .common import (
    BookingMode,
    Currency,
    MercuryModel,
    OverCapacityPolicy,
    SoftDeleteMixin,
    TimestampMixin,
)


class Service(MercuryModel, TimestampMixin, SoftDeleteMixin):
    """A bookable service offered by an establishment.

    Booking mode is per-service (ADR-0004), not per-establishment.
    """

    id: str | None = None

    # Relationships (record IDs as strings — resolved by repository layer)
    establishment: str = ""
    group: str | None = None  # record<service_group>
    assigned_staff: list[str] = Field(default_factory=list)  # array<record<staff>>
    required_resource_groups: list[str] = Field(default_factory=list)

    # Identity
    title: str = ""
    description: str | None = None

    # Booking mode (ADR-0004)
    booking_mode: BookingMode = BookingMode.STANDARD

    # Media
    cover_image: str | None = None
    gallery: list[str] | None = None

    # Taxonomy
    service_type: list[str] = Field(default_factory=list)
    subcategory: str | None = None

    # AI / Discovery
    keywords: list[str] = Field(default_factory=list)
    ai_description: str | None = None
    embedding: list[float] | None = None

    # Time & Pricing
    duration: int = 30  # minutes
    price: float = 0.0
    buffer_time: int = 0  # minutes
    currency: Currency = Currency.NOK

    # Policies
    over_capacity_policy: OverCapacityPolicy = OverCapacityPolicy.REJECT
    min_booking_notice_minutes: int = 0
    slot_interval: int = 15

    # Status
    is_active: bool = True


class ServiceGroup(MercuryModel, TimestampMixin, SoftDeleteMixin):
    """Logical grouping of services for display and booking panel.

    Source: company-blueprint.surql §1.3
    """

    id: str | None = None
    establishment: str = ""
    name: str = ""
    description: str | None = None
    staff_ids: list[str] | None = None
    sort_order: int = 0
    show_on_booking_panel: bool = True
    multi_select: bool = False
